infra/build/functions/update_build_status.py

```diff
@@ -24,71 +24,138 @@
+          'finish_time': project_build['finishTime'],
+      }
+
     if not builds_status.upload_log(build_id):
       log_name = 'log-{0}'.format(build_id)
       raise MissingBuildLogError('Missing build log file {0}'.format(log_name))
 
-    return project_build
+    history.append({
+        'build_id': build_id,
+        'finish_time': project_build['finishTime'],
+        'success': builds_status.is_build_successful(project_build)
+    })
+
+    if len(history) == MAX_BUILD_LOGS:
+      break
 
-  return None
+  project = {'history': history}
+  if last_successful_build:
+    project['last_successful_build'] = last_successful_build
+  return project
 
 
+# pylint: disable=too-many-locals
 def update_build_status(build_tag, status_filename):
   """Update build statuses."""
+  projects = []
   statuses = {}
-  successes = []
-  failures = []
   for project_build in BuildsHistory.query(
-      BuildsHistory.build_tag == build_tag):
-    last_build = get_last_build(project_build.build_ids)
-    if not last_build:
-      logging.error('Failed to get last build for project %s',
-                    project_build.project)
+      BuildsHistory.build_tag == build_tag).order('project'):
+
+    project = get_build_history(project_build.build_ids)
+    project['name'] = project_build.project
+    projects.append(project)
+    last_successful_build = ndb.Key(LastSuccessfulBuild,
+                                    project_build.name + '-' + build_tag).get()
+    if project['history']:
+      statuses[project_build.name] = project['history'][0]['success']
+
+    if not last_successful_build and 'last_successful_build' not in project:
       continue
 
-    if last_build['status'] == 'SUCCESS':
-      statuses[project_build.project] = True
-      successes.append({
-          'name': project_build.project,
-          'build_id': last_build['id'],
-          'finish_time': last_build['finishTime'],
-          'success': True,
-      })
+    if 'last_successful_build' not in project:
+      project['last_successful_build'] = {
+          'build_id': last_successful_build.build_id,
+          'finish_time': last_successful_build.finish_time
+      }
     else:
-      statuses[project_build.project] = False
-      failures.append({
-          'name': project_build.project,
-          'build_id': last_build['id'],
-          'finish_time': last_build['finishTime'],
-          'success': False,
-      })
-
-  builds_status.upload_status(successes, failures, status_filename)
+      if last_successful_build:
+        last_successful_build.build_id = project['last_successful_build'][
+            'build_id']
+        last_successful_build.finish_time = project['last_successful_build'][
+            'finish_time']
+      else:
+        last_successful_build = LastSuccessfulBuild(
+            id=project_build.name + '-' + build_tag,
+            project=project_build.name,
+            uild_id=project['last_successful_build']['build_id'],
```

Comment:
looks like a typo here in uild_id. Did your tests not catch this?

Reply:
I hadn't covered everything last night, had covered get_build_history and sort_projects, will add a few unit tests for this as well now